File: src/test/java/toby/command/commands/misc/UserInfoCommandTest.java
package toby.command.commands.misc;

import commands.CommandTest;
import net.dv8tion.jda.api.entities.Guild;
import net.dv8tion.jda.api.entities.Member;
import net.dv8tion.jda.api.entities.Mentions;
import net.dv8tion.jda.api.interactions.commands.OptionMapping;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.mockito.Mock;
import toby.command.CommandContext;
import toby.jpa.dto.MusicDto;
import toby.jpa.dto.UserDto;
import toby.jpa.service.IUserService;

import java.util.List;

import static org.mockito.ArgumentMatchers.anyLong;
import static org.mockito.Mockito.*;

public class UserInfoCommandTest implements CommandTest {

    @Mock
    private IUserService userService;

    private UserInfoCommand userInfoCommand;

    @BeforeEach
    public void setUp() {
        setUpCommonMocks(); // Call the common setup defined in the interface

        // Additional setup specific to UserInfoCommandTest
        userService = mock(IUserService.class);
        userInfoCommand = new UserInfoCommand(userService);
        when(event.getGuild()).thenReturn(mock(Guild.class));
    }

    @Test
    public void testHandleCommandWithOwnUser() {

        // Mock the event's options to be empty
        when(event.getOptions()).thenReturn(List.of());

        // Mock the requesting user's DTO
        UserDto requestingUserDto = new UserDto();
        when(userService.getUserById(anyLong(), anyLong())).thenReturn(requestingUserDto);
        requestingUserDto.setMusicDto(new MusicDto());
        when(event.getGuild().getIdLong()).thenReturn(123L);

        // Test handle method
        userInfoCommand.handle(new CommandContext(event), requestingUserDto, 0);

        // Verify interactions
        verify(interactionHook, times(1)).sendMessageFormat(anyString(), eq(requestingUserDto));
        verify(interactionHook, times(1)).sendMessage(anyString());
        verify(userService, times(0)).getUserById(anyLong(), anyLong());
    }

    @Test
    public void testHandleCommandWithMentionedUserAndValidRequestingPermissions() {
        // Mock user interaction with mentioned user

        // Mock the event's options to include mentions
        when(event.getOptions()).thenReturn(List.of(mock(OptionMapping.class))); // Add an option to simulate mentions

        // Mock a mentioned user's DTO
        UserDto mentionedUserDto = new UserDto();
        when(userService.getUserById(anyLong(), anyLong())).thenReturn(mentionedUserDto);
        mentionedUserDto.setMusicDto(new MusicDto());

        // Mock mentions
        when(event.getOption(anyString())).thenReturn(mock(OptionMapping.class));
        Mentions mentions = mock(Mentions.class);
        when(event.getOption("users").getMentions()).thenReturn(mentions);
        Member mockMember = mock(Member.class);
        List<Member> memberList = List.of(mockMember);
        when(mentions.getMembers()).thenReturn(memberList);
        Guild mockGuild = mock(Guild.class);
        when(mockMember.getGuild()).thenReturn(mockGuild);
        when(mockGuild.getIdLong()).thenReturn(123L);
        UserDto mentionedUserMock = mock(UserDto.class);
        when(userService.getUserById(anyLong(), anyLong())).thenReturn(mentionedUserMock);
        when(mentionedUserMock.getMusicDto()).thenReturn(new MusicDto(1L, 1L, "filename", 10, null));
        when(mockMember.getEffectiveName()).thenReturn("Toby");


        // Test handle method
        UserDto userDto = mock(UserDto.class);
        when(userDto.isSuperUser()).thenReturn(true);
        userInfoCommand.handle(new CommandContext(event), userDto, 0);

        // Verify interactions

        //lookup on mentionedMember
        verify(userService, times(1)).getUserById(anyLong(), anyLong());
        //music file message
        verify(interactionHook, times(1)).sendMessageFormat(anyString(), anyString());
        //mentioned user message
        verify(interactionHook, times(1)).sendMessageFormat(anyString(), eq("Toby"), eq(mentionedUserMock));
    }

    @Test
    public void testHandleCommandNoPermission() {
        // Mock user interaction without permission

        // Mock the event's options to include mentions
        when(event.getOptions()).thenReturn(List.of(mock(OptionMapping.class)));

        // Mock the requesting user without permission
        UserDto requestingUserDto = mock(UserDto.class);
        when(requestingUserDto.isSuperUser()).thenReturn(false);

        // Test handle method
        userInfoCommand.handle(new CommandContext(event), requestingUserDto, 0);

        // Verify interactions
        verify(interactionHook, times(1)).sendMessage(anyString());
    }
}

File: src/test/kotlin/toby/command/commands/misc/UserInfoCommandTest.kt
package toby.command.commands.misc

import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import net.dv8tion.jda.api.entities.Mentions
import net.dv8tion.jda.api.interactions.commands.OptionMapping
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import toby.command.CommandContext
import toby.command.CommandTest
import toby.command.CommandTest.Companion.event
import toby.command.CommandTest.Companion.member
import toby.command.CommandTest.Companion.requestingUserDto
import toby.jpa.dto.MusicDto
import toby.jpa.dto.UserDto
import toby.jpa.service.IUserService

class UserInfoCommandTest : CommandTest {
    lateinit var userService: IUserService
    lateinit var userInfoCommand: UserInfoCommand

    @BeforeEach
    fun setUp() {
        setUpCommonMocks() // Call the common setup defined in the interface

        // Additional setup specific to UserInfoCommandTest
        userService = mockk()
        userInfoCommand = UserInfoCommand(userService)
    }

    @AfterEach
    fun tearDown() {
        tearDownCommonMocks()
    }

    @Test
    fun testHandleCommandWithOwnUser() {
        // Mock the event's options to be empty
        every { event.options } returns listOf()

        // Mock the requesting user's DTO
        val userDto = UserDto(
            1, 1,
            superUser = true,
            musicPermission = true,
            digPermission = true,
            memePermission = true,
            socialCredit = 0,
            initiativeModifier = 0,
            musicDtos = emptyList<MusicDto>().toMutableList()
        )
        every { userService.getUserById(any(), any()) } returns userDto
        requestingUserDto.apply {
            every { musicDtos } returns listOf(MusicDto()).toMutableList()
        }

        // Test handle method
        userInfoCommand.handle(CommandContext(event), userDto, 0)

        // Verify interactions
        verify(exactly = 1) {
            event.hook.sendMessage("Here are the permissions for 'Effective Name': 'MUSIC: true, MEME: true, DIG: true, SUPERUSER: true'. \n" +
                    " There is no valid intro music file associated with user Effective Name.")
        }
        verify(exactly = 1) {
            userService.getUserById(any(), any())
        }
    }

    @Test
    fun testHandleCommandWithMentionedUserAndValidRequestingPermissions() {
        // Mock the requesting user's DTO
        val userDto = UserDto(
            1, 1,
            superUser = true,
            musicPermission = true,
            digPermission = true,
            memePermission = true,
            socialCredit = 0,
            initiativeModifier = 0,
            musicDtos = emptyList<MusicDto>().toMutableList()
        )

        // Mock the event's options to include mentions
        every { event.options } returns listOf(mockk<OptionMapping>())

        // Mock a mentioned user's DTO
        val mentionedUserDto = UserDto(6L, 1L)
        every { userService.getUserById(any(), any()) } returns userDto
        every { requestingUserDto.musicDtos } returns listOf(
            MusicDto(
                UserDto(1, 1),
                1,
                "filename",
                10,
                null
            )
        ).toMutableList()

        // Mock mentions
        val mentions = mockk<Mentions>()
        every { event.getOption(any())?.mentions } returns mentions
        val memberList = listOf(member)
        every { mentions.members } returns memberList
        every { userService.getUserById(any(), any()) } returns mentionedUserDto

        // Test handle method
        userInfoCommand.handle(CommandContext(event), requestingUserDto, 0)

        // Verify interactions
        verify(exactly = 1)
        {
            userService.getUserById(any(), any())
            event.hook.sendMessage(any<String>())
        }
    }

    @Test
    fun testHandleCommandNoPermission() {
        // Mock user interaction without permission
        every { userService.getUserById(any(), any()) } returns requestingUserDto
        every { requestingUserDto.superUser } returns false

        // Mock the event's options to include mentions
        every { event.options } returns listOf(mockk<OptionMapping>())

        // Mock the requesting user without permission
        // Test handle method
        userInfoCommand.handle(CommandContext(event), requestingUserDto, 0)

        // Verify interactions
        verify(exactly = 1) {
            event.hook.sendMessage("You do not have permission to view user permissions, if this is a mistake talk to the server owner")
        }
    }
}